core/trino-main/src/main/java/io/trino/split/BufferingSplitSource.java

```diff
@@ -72,50 +78,71 @@ public Optional<List<Object>> getTableExecuteSplitsInfo()
     }
 
     private static class GetNextBatch
+            extends AbstractFuture<SplitBatch>
     {
         private final Context context = Context.current();
         private final SplitSource splitSource;
+        private final Executor executor;
         private final int min;
         private final int max;
 
+        @GuardedBy("this")
         private final List<Split> splits = new ArrayList<>();
-        private boolean noMoreSplits;
+        @GuardedBy("this")
+        private ListenableFuture<SplitBatch> nextBatchFuture;
 
         public static ListenableFuture<SplitBatch> fetchNextBatchAsync(
                 SplitSource splitSource,
+                Executor executor,
                 int min,
                 int max)
         {
-            GetNextBatch getNextBatch = new GetNextBatch(splitSource, min, max);
-            ListenableFuture<Void> future = getNextBatch.fetchSplits();
-            return Futures.transform(future, ignored -> new SplitBatch(getNextBatch.splits, getNextBatch.noMoreSplits), directExecutor());
+            GetNextBatch getNextBatch = new GetNextBatch(splitSource, executor, min, max);
+            getNextBatch.fetchSplits();
+            return getNextBatch;
         }
 
-        private GetNextBatch(SplitSource splitSource, int min, int max)
+        private GetNextBatch(SplitSource splitSource, Executor executor, int min, int max)
         {
             this.splitSource = requireNonNull(splitSource, "splitSource is null");
+            this.executor = requireNonNull(executor, "executor is null");
             checkArgument(min <= max, "Min splits greater than max splits");
             this.min = min;
             this.max = max;
         }
 
-        private ListenableFuture<Void> fetchSplits()
+        private synchronized void fetchSplits()
         {
-            if (splits.size() >= min) {
-                return immediateVoidFuture();
-            }
-            ListenableFuture<SplitBatch> future;
+            verify(nextBatchFuture == null || nextBatchFuture.isDone(), "getNextBatch should not be already running");
+
             try (var ignored = context.makeCurrent()) {
-                future = splitSource.getNextBatch(max - splits.size());
+                nextBatchFuture = splitSource.getNextBatch(max - splits.size());
```

Comment:
In the situation where the returned future is already done, we probably don't want to push back into an async context, but would rather just return from the current thread directly.

Reply:
That depends of number of returner splits. If you use `directExecutor` it might still cause stack overflow

Reply:
Correct, but it's not strictly necessary to use a `directExecutor`. If you just handle the finished future's value directly then you stay at the same call stack height and you can either return an immediate future or call back into the connector for the next batch from there without creating unnecessary thread creation activity through the new executor.

Reply:
I've made changes now to handle completed futures directly without using separate executor or recursion